src/pubsub/ua_pubsub_reader.c

```diff
@@ -151,11 +392,259 @@ UA_StatusCode
+                    break;
+                case UA_PUBSUBSTATE_PAUSED:
+                    break;
+                case UA_PUBSUBSTATE_OPERATIONAL:
+                    return UA_STATUSCODE_GOOD;
+                case UA_PUBSUBSTATE_ERROR:
+                    break;
+                default:
+                    UA_LOG_WARNING(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                                   "Received unknown PubSub state!");
+            }
+            break;
+        case UA_PUBSUBSTATE_ERROR:
+            switch (readerGroup->state){
+                case UA_PUBSUBSTATE_DISABLED:
+                    break;
+                case UA_PUBSUBSTATE_PAUSED:
+                    break;
+                case UA_PUBSUBSTATE_OPERATIONAL:
+                    break;
+                case UA_PUBSUBSTATE_ERROR:
+                    return UA_STATUSCODE_GOOD;
+                default:
+                    UA_LOG_WARNING(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                                   "Received unknown PubSub state!");
+            }
+            break;
+        default:
+            UA_LOG_WARNING(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                           "Received unknown PubSub state!");
+    }
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_StatusCode
+UA_Server_freezeReaderGroupConfiguration(UA_Server *server, const UA_NodeId readerGroupId) {
+    UA_ReaderGroup *rg = UA_ReaderGroup_findRGbyId(server, readerGroupId);
+    if(!rg)
+        return UA_STATUSCODE_BADNOTFOUND;
+
+    //PubSubConnection freezeCounter++
+    UA_NodeId pubSubConnectionId =  rg->linkedConnection;;
+    UA_PubSubConnection *pubSubConnection = UA_PubSubConnection_findConnectionbyId(server, pubSubConnectionId);
+    pubSubConnection->configurationFreezeCounter++;
+    pubSubConnection->configurationFrozen = UA_TRUE;
+    //ReaderGroup freeze
+    rg->configurationFrozen = UA_TRUE;
+    // TODO: Clarify on the freeze functionality in multiple DSR, multiple networkMessage conf in a RG
+    //DataSetReader freeze
+    UA_DataSetReader *dataSetReader;
+    UA_UInt16 dsrCount = 0;
+    LIST_FOREACH(dataSetReader, &rg->readers, listEntry){
+    	dataSetReader->configurationFrozen = UA_TRUE;
+        dsrCount++;
+        /* TODO: Configuration frozen for subscribedDataSet once
+         * UA_Server_DataSetReader_addTargetVariables API modified to support
+         * adding target variable one by one or in a group stored in a list.
+         */
+    }
+
+    if(rg->config.rtLevel == UA_PUBSUB_RT_FIXED_SIZE) {
+        if(dsrCount > 1) {
+            UA_LOG_WARNING(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                           "Mutiple DSR in a readerGroup not supported in RT fixed size configuration");
+            return UA_STATUSCODE_BADNOTIMPLEMENTED;
+        }
+
+        dataSetReader = LIST_FIRST(&rg->readers);
+        // TODO: Check for UADP encoding
```

Comment:
Can we add this check within this PR?

Reply:
Done. Modified to Extension object and added the checks required in a separate commit.